### PR TITLE
fix inconsistent tabularInput keys

### PR DESCRIPTION
**Pull request by yeukfu**
Fix the issue (https://github.com/rlworkgroup/dowel/issues/45) by rewrite csv file when new key appears. The performance may be not good when the existing csv file is large, as we need to rewrite all the records. But every record is persistent in this way.
To leave the cell blank when the value of a key missing, TabularInput.clear() need to call after logger.dump_all().

### CODE REVIEW

File: src/dowel/csv_output.py [hunk 1]
```diff
@@ -41,13 +41,22 @@ def record(self, data, prefix=''):
...
+                if key not in self._fieldnames:
+                    self._fieldnames.add(key)
+                    new_key_occur = True
+
+            if new_key_occur:
+                reader = csv.DictReader(self._log_file_read)
+                self._writer = csv.DictWriter(
+                    self._log_file,
+                    fieldnames=self._fieldnames,
+                    extrasaction='ignore')
```
Comment (krzentner) [inline src/dowel/csv_output.py:55]: Why use `extrasaction='ignore'` here? It seems like `extrasaction='raise'` would be a safer choice.

File: src/dowel/simple_outputs.py [hunk 2]
```diff
@@ -60,6 +60,7 @@ def __init__(self, file_name, mode='w'):
         mkdir_p(os.path.dirname(file_name))
         # Open the log file in child class
         self._log_file = open(file_name, mode)
+        self._log_file_read = open(file_name, 'r')
```
[the remark this answers is not in the document]
Comment (yeukfu) [reply src/dowel/simple_outputs.py]: To deal with the undefine behavior. In my new PR, I open the file once in w+ mode.

File: src/dowel/csv_output.py [hunk 3]
```diff
@@ -41,13 +41,22 @@ def record(self, data, prefix=''):
...
-                               set(self._fieldnames), set(to_csv.keys())))
+            new_key_occur = False 
+            for key in to_csv.keys():
+                if key not in self._fieldnames:
+                    self._fieldnames.add(key)
+                    new_key_occur = True
+
+            if new_key_occur:
+                reader = csv.DictReader(self._log_file_read)
+                self._writer = csv.DictWriter(
```
[the remark this answers is not in the document]
Comment (yeukfu) [reply src/dowel/csv_output.py:49]: I refractor out the writer initialization. However, in order to use csv.DictWriter, if the parameters change, it is unavoidable to reinitiallize writer.

File: src/dowel/csv_output.py (hunk 1 above)
Comment (yeukfu) [reply src/dowel/csv_output.py]: Thank you for point out it. extrasaction='raise' will be better

**Comment by yeukfu** [comment]
As open the same file twice is not well defined, I open the file only once in w+ mode to read and write the csv file.

**Pull request closed without merging**